---
title: 'TCR Analysis'
author: |
 | Derrik Gratz
 | FHIL
date: '`r Sys.Date()`'
output: 
  rmdformats::robobook:
    fig_width: 9
    toc_depth: 3
    css: "robobook.css"
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE,
                      fig.width = 6, fig.height = 8, dev='svglite',
                      cache=FALSE) 
library(kableExtra)
```

# Overview

This report shows data from the bulk TCR sequencing from the BM03 project processed using [MixCR](https://mixcr.com/). 
MixCR settings were kept as consistent as possible. Primer locations and UMI descriptions were changed based on specifications in kit manuals. 
Data was generated with the following kits:

[Qiaseq Targeted RNA-seq Panel for T-cell receptor kit](https://www.qiagen.com/us/products/discovery-and-translational-research/next-generation-sequencing/rna-sequencing/t-cell-receptor-sequencing/qiaseq-targeted-rnaseq-panel-for-t-cell-receptor)
[Takara SMARTer Human TCR a/b Profiling Kit v2](https://www.takarabio.com/products/next-generation-sequencing/immune-profiling/human-repertoire/human-tcrv2-profiling-kit-for-illumina-sequencing)
[NEBNext Immune Sequencing Kit (Human)](https://www.neb.com/en-us/products/e6320-nebnext-immune-sequencing-kit-human#Product%20Information)
[Cellecta DriverMap AIR TCR-BCR Profiling Kit (Human RNA)](https://cellecta.com/collections/drivermap-adaptive-immune-receptor-air-profiling/products/drivermap%E2%84%A2-air-tcr-bcr-profiling-kit-human-rna-24-and-96-rxn?variant=43268706435330)

Samples follow a naming convention describing 

* The kit used
* The source sample ID
* The input RNA amount (L=large, s=small)
* The replicate ID (if applicable, not every sample had replicates)

```{r}
knitr::kable(qc_metadata %>% select(-chain))
```


***

# QC

<details><summary>Full explanation of metrics</summary>

[Source: MixCR](https://mixcr.com/mixcr/reference/qc-list-of-checks/)

**Successfully aligned reads:**

Sequencing reads that have been accurately mapped to their respective V, D, and J gene segments during the alignment process. This metric is crucial for assessing data quality. Low rate of aligned reads is a strong signal of either wet lab problems or misused analysis settings.

**Off target (non TCR/IG) reads:**

Sequencing reads that do not map on T or B cell receptor gene segments. High rate of off-target reads may be a result of:

* primers mis-annealing to a non-target loci;
* DNA contamination in RNA material;
* other protocol and/or wet lab issues;
* wrong choice of species in analysis settings (e.g. you have a cat library while used a human reference library).

In order to troubleshoot, one can rerun alignment and save not aligned reads into separate files (using --not-aligned-R1 and --not-aligned-R2 options). Then pick a few random non-aligned sequences and BLAST them to identify whether they are coming from contamination / non-target loci or aligned to different species. If there are no BLAST hits, check whether the sequences are artificial (e.g. adapters).

**Reads with no V or J hits:**

Sequencing reads that cover either V or J region but not both of them simultaneously. There might be several reasons for that. One common reason is incorrect orientation of reads, typically caused by some pre-processing of the input data performed before MiXCR. A common example is pre-processing with external tools like MiGEC, which is a legacy tool for handling UMIs and demultiplexing. MiGEC reverse complements one of the reads, thus requiring running MiXCR with -OreadsLayout=Collinear option.

Another reason is a very low sequencing quality in one of the reads when sequencing has completely failed.

Finally, it might be a wrong use of the analysis preset: if the input data is randomly fragmented (RNA-, Exome-Seq, 10x etc.) and the used preset is designed for amplicon libraries, you would see a high percentage here (since with amplicon settings MiXCR drops all reads that do not full cover CDR3 region). So check and use an appropriate preset for fragmented data.
Reads with no barcode

For barcoded data means that barcodes can't be extracted using the specified tag pattern. For custom library preparation protocols, the first thing to check is that a correct tag pattern and/or a correct analysis preset are used. If the library is unstranded, one should use either --tag-parse-unstranded option or change the preset accordingly. Finally, the quality of data may be low, so that tags can't be parsed because of too many sequencing errors.

**Reads dropped in $tag error correction and whitelist:**

Sequencing reads that were dropped along with dropped bad quality barcodes during the UMI sequencing and PCR error correction. The high rate of dropped reads here signals about poor sequencing quality and related issues.
Reads dropped in tags filtering

Sequencing reads that were dropped in barcode filtering. High rate of dropped reads may signal either about poor quality of the library, significant under sequencing, mistakes at library pooling and various other reasons.
Tag groups with no assembling feature

Number of tag groups which do not contain any reads that cover clonotype assembling feature. For amplicon protocols one should expect around zero tag groups with no assembling feature. A high rate here signals about either failed wet lab library preparation or misuse of the analysis settings. A typical scenario of wrong settings is when one specifies full-length VDJRegion as an assembling feature, while either the protocol is not designed for this purpose (i.e. multiplex protocol with primers in frameworks) or short sequencing was used (e.g. 150+150).

**Barcode collisions in clonotype assembly:**

Number of barcodes that resulted in multiple consensus sequences. Typically, one expects that one tag group (e.g. one UMI) results in exactly a single consensus sequence, so that the sequences with the same UMI are highly similar (within several mismatches or indels coming from PCR and sequencing errors). A tiny fraction of barcodes may result in more than one consensus due to the birth paradox. High collision rate signals about poor barcode diversity related to e.g. too short barcode sequence, protocol design problems or some wet lab issues.

**Unassigned alignments in clonotype assembly:**

Number of alignments that we dropped in clonotype assembly because of the high barcode collision rate. Typically, one expects that the sequences with the same tag group (e.g. UMI) are coming from the same molecule and are highly similar (within several mismatches or indels coming from PCR and sequencing errors). When a single tag group has too many reads with completely different CDR3 sequences, MiXCR will drop them as unreliable. A high rate of unassigned alignments may signal about suboptimal design of the library architecture and preparation protocol. Manual inspection of raw reads will help to debug the exact problem.

**Reads used in clonotypes:**

Percent of the raw sequencing reads used in the resulting clonotypes. For high quality of data prepared with targeted VDJ protocols (either amplicon or fragmented), one expects that a significant percentage of raw sequencing reads are finally used to build clonotypes. Low percent of used clonotypes may signal about various problems, and other QC metrics may help to understand the reason.

**Alignments without assembling feature:**

Number of aligned reads that do not cover clonotype assembling feature. Each amplicon protocol is designed to cover a specific part of the VDJ region, depending on the position of primers. High rate of alignments which do not cover assembling feature specified signals about wrong setting for the assembling feature. The most common reason is that one is trying to extract full-length clonal sequences of the receptors (VDJRegion), but the library is prepared in such a way that it does not cover the full length of the receptor or a short sequencing was used (to cover the full length one at least need 250+250 bp technology).

**Alignments dropped due to low sequence quality:**

Number of aligned sequencing reads that contain low sequencing quality nucleotides inside clonotype assembling feature (e.g. CDR3) and that MiXCR was unable to rescue by assigning to high quality clonotypes in error correction. High rate of such alignments is a clear signal of poor sequencing quality (check sequencing quality reports).

**Alignments clustered in PCR error correction:**

Number of aligned reads used in the clonotypes that were clustered in PCR error correction. Since the background PCR error rate is more or less same for all amplicon protocols, a high rate here would be a signal of misused analysis setting in most cases (for example use of an inappropriate reference library).

**Clonotypes clustered in PCR error correction:**

Number of clonotypes that were clustered in PCR error correction. Since the background PCR error rate is more or less same for all amplicon protocols, a high rate here would be a signal of misused analysis setting in most cases (for example use of a inappropriate reference library).

**Clones dropped in post-filtering:**

Number of clonotypes that were dropped in the post filtering.

**Alignments dropped in clones post-filtering:**

Number of aligned reads that were dropped along with clonotypes in the post filtering.

</details>

## Small 

```{r, fig.width = 10, fig.height = 8}
figures[['QC_small']]
```

## Large

```{r, fig.width = 10, fig.height = 8}
figures[['QC_large']]
```

***

# Clone recovery

Number of recovered chains/clones

## Large {.tabset}

### TRA/TRB

```{r, fig.width = 6, fig.height = 4}
figures[['chain_recovery_large_TRATRB']]
```

### Other

```{r}
figures[['chain_recovery_large_other']]
```


## Small {.tabset}

### TRA/TRB

```{r, fig.width = 6, fig.height = 4}
figures[['chain_recovery_small_TRATRB']]
```

### Other

```{r, fig.width = 6, fig.height = 4}
figures[['chain_recovery_small_other']]
```

***

# Unique clone count

Number of unique chains/clones

## Large {.tabset}

### TRATRB

```{r, fig.width = 6, fig.height = 4}
figures[['unique_clones_large_TRATRB']]
```

### other

```{r, fig.width = 6, fig.height = 4}
figures[['unique_clones_large_other']]
```

## Small {.tabset}

### TRATRB

```{r, fig.width = 6, fig.height = 4}
figures[['unique_clones_small_TRATRB']]
```

### other

```{r, fig.width = 6, fig.height = 4}
figures[['unique_clones_large_other']]
```

***

# Clone count proportions

Relative abundance for the groups of rare clonotypes with low counts

## Large

```{r, fig.width = 8, fig.height = 4}
figures[['clone_count_proportion_large_TRATRB']]
```

## Small

```{r, fig.width = 8, fig.height = 4}
figures[['clone_count_proportion_small_TRATRB']]
```

***

# Clone frequency proportion

Proportion of repertoire occupied by clonal groups with specific abundances.

## Large

```{r, fig.width = 8, fig.height = 4}
figures[['clone_frequency_proportion_large_TRATRB']]
```

## Small

```{r, fig.width = 8, fig.height = 4}
figures[['clone_frequency_proportion_small_TRATRB']]
```

***

# Diversity estimation

Inverse Simpson index - effective number of types that is obtained when the weighted arithmetic mean is used to quantify average proportional abundance of types in the dataset of interest.

## Large

```{r, fig.width = 6, fig.height = 4}
figures[['simpson_large']]
```

## Small

```{r, fig.width = 6, fig.height = 4}
figures[['simpson_small']]
```

***

# Clonal overlap

Measuring similarity between samples by overlap of repertoires

## All {.tabset}

### TRA

```{r, fig.width=8, fig.height=8}
figures[['overlap_all_tra']] 
```

### TRB

```{r, fig.width=8, fig.height=8}
figures[['overlap_all_trb']] 
```

## Small {.tabset}

### TRA

```{r, fig.width=8, fig.height=8}
figures[['overlap_small_tra']] 
```

### TRB

```{r, fig.width=8, fig.height=8}
figures[['overlap_small_trb']] 
```

## Large {.tabset}

### TRA

```{r, fig.width=8, fig.height=8}
figures[['overlap_large_tra']] 
```

### TRB

```{r, fig.width=8, fig.height=8}
figures[['overlap_large_trb']] 
```

***

# Saturation curves

Number of unique clonotypes observed at different downsampling rates. This is slightly different than a traditional *sequencing* saturation curve relating UMIs to total reads. This relationship addresses the question: **what target UMIs/cell do we need to capture the full repertoire of the input?** Downsampling was performed on the UMI level, *after* processing, which may be slightly biased, but is much more computationally efficient than downsampling and rerunning the full processing pipeline, but that could be done if preferred.

## Large

```{r}
figures[['clone_saturation_curves_large']]
```

## Small

```{r}
figures[['clone_saturation_curves_small']]
```
