---
title: 'TCR QC'
author: "FHIL\nDerrikGratz"
date: '`r Sys.Date()`'
output: html_document
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      fig.width = 8, fig.height = 6,
                      cache=FALSE, cache.lazy = FALSE, dev='svg') 

library(tidyverse)  ## General R logic
library(here)       ## Easier specification of file locations
library(ggbeeswarm) ## swarm plots
library(stringr)    ## Str_match
library(data.table) ## rbindlist
set.seed(44)
```

```{r}
figures <- list()
```

# Functions

```{r}
parse_name <- function(x) {
  x <- as.data.frame(str_match(x, '(^.)-(F\\d)(L|S)(A|B)?_?(.+)?'))
  colnames(x) <- c(
    'sample', 'kit', 'individual', 'size', 'replicate', 'chain'
  )
  x %>%
    mutate(
      replicate = if_else(is.na(replicate), 'A', replicate),
      kit = case_when(
        kit == 'N' ~ 'NEB',
        kit == 'Q' ~ 'QiaSeq',
        kit == 'T' ~ 'Takara',
        .default = 'unknown'
      )
    )
}
```

# Pipeline QC

```{r}
data_dir <- here('../processing/mixcr/downsampled_data/by_umi/')
qc_metrics <- list.files(path = data_dir, pattern = '*QC.txt', recursive = TRUE, full.names = TRUE)
names(qc_metrics) <- gsub('.+(.-F\\w+)_QC.+', '\\1', qc_metrics)
qc_metrics <- lapply(qc_metrics, read.table, sep=':', strip.white=TRUE)
qc_metrics <- data.table::rbindlist(qc_metrics, idcol='sample')
qc_metrics <- qc_metrics %>%
  separate(V2, c('value', 'status'), sep = '\\[', ) %>%
  mutate(value = readr::parse_number(value)) %>%
  mutate(status = substr(status, 1, nchar(status)-1)) %>%
  mutate(status = factor(status, levels = c('OK', 'WARN', 'ALERT')))
```

```{r}
qc_metadata <- rbindlist(lapply(unique(qc_metrics$sample), parse_name))
qc_metadata <- qc_metadata %>%
  mutate(across(everything(),
                ~ factor(., levels = gtools::mixedsort(unique(.)))
                ))
```

```{r, fig.width=10, fig.height=8}
metrics <- c(
  "Off target (non TCR/IG) reads", 
  "Reads with no V or J hits",
  "Reads used in clonotypes", 
  "Successfully aligned reads",
  "Alignments that do not cover CDR3", 
  "Unassigned alignments in clonotype assembly" ,
  "UMIs artificial diversity eliminated", 
  "Reads dropped in tags error correction and filtering",
  "Barcode collisions in clonotype assembly"
)
figures[['QC_large']] <-
  left_join(qc_metrics, qc_metadata, by='sample') %>%
  filter(grepl('L', sample)) %>%
  filter(V1 %in% metrics) %>% 
  mutate(V1 = factor(V1, levels=metrics)) %>%
ggplot(aes(x=kit, y=value)) +
  geom_boxplot(outliers = FALSE) +
  geom_beeswarm(aes(color=status, shape=individual), size=2) +
  theme_bw() + 
  scale_color_manual(values = c("OK"='#34C400', 'WARN'='goldenrod', 'ALERT'='#F54411')) +
  # theme(text = element_text(family='Courier', size = 11)) +
  # theme(axis.title.x=element_blank(),
  #       axis.text.x=element_blank(),
  #       axis.ticks.x=element_blank()) +
  facet_wrap(~V1, scales = 'free', labeller = label_wrap_gen(40), ncol=3) +
  labs(x='Sample', y='Value (%)', color='MixCR\nstatus')

figures[['QC_large']]
```

```{r, fig.width=10, fig.height=8}
metrics <- c(
  "Off target (non TCR/IG) reads", 
  "Reads with no V or J hits",
  "Reads used in clonotypes", 
  "Successfully aligned reads",
  "Alignments that do not cover CDR3", 
  "Unassigned alignments in clonotype assembly" ,
  "UMIs artificial diversity eliminated", 
  "Reads dropped in tags error correction and filtering",
  "Barcode collisions in clonotype assembly"
)
figures[['QC_small']] <-
  left_join(qc_metrics, qc_metadata, by='sample') %>%
  filter(grepl('S', sample)) %>%
  filter(V1 %in% metrics) %>% 
  mutate(V1 = factor(V1, levels=metrics)) %>%
ggplot(aes(x=kit, y=value)) +
  geom_boxplot(outliers = FALSE) +
  geom_beeswarm(aes(color=status, shape=individual), size=2) +
  theme_bw() + 
  scale_color_manual(values = c("OK"='#34C400', 'WARN'='goldenrod', 'ALERT'='#F54411')) +
  # theme(text = element_text(family='Courier', size = 11)) +
  # theme(axis.title.x=element_blank(),
  #       axis.text.x=element_blank(),
  #       axis.ticks.x=element_blank()) +
  facet_wrap(~V1, scales = 'free', labeller = label_wrap_gen(40), ncol=3) +
  labs(x='Sample', y='Value (%)', color='MixCR\nstatus')

figures[['QC_small']]
```


# FastQC

```{r}
data_dir <- dir(here('../processing/fastqc/'), 'fastqc_data.txt', recursive = TRUE, full.names = TRUE)
fastqc <- lapply(data_dir, 
       function(x) {
         sample <- gsub('.*(.-F\\d[^_]+).+', '\\1', x)
         x <- readLines(x,n=8)[c(7,8)]
         x <- strsplit(x, '\t')
         return(list(sample=sample, total_sequences=x[[1]][2], poor_sequences=x[[2]][2]))
}) %>%
  rbindlist() %>%
  unique()
```

```{r}
for (i in c('L', 'S')) {
  for (k in c('NEB', 'Takara', 'QiaSeq')) {
    figures[[paste0(c('fastqc', k, i), collapse ='_')]] <- 
      fastqc %>%
      left_join(qc_metadata, by = 'sample') %>%
      filter(size==i,
             kit==k) %>%
      select(sample, total_sequences, poor_sequences) %>%
      mutate(total_sequences = as.numeric(total_sequences),
             poor_sequences = as.numeric(poor_sequences)) %>%
      knitr::kable(format.args = list(big.mark = ","))
  }
}
figures[['fastqc_NEB_S']]
```

# Saturation curves

## UMIs

```{r}
files <- dir(here('../processing/mixcr/'), 
             'downsampling_summary.*.txt', 
             recursive = TRUE, 
             full.names = TRUE)
downsampling_umis <- lapply(
  files, 
  function(file) {
    sample <- gsub('.*(.-F[0-9][^/]+).*', '\\1', file)
    size <- gsub('.*_summary_([0-9]*).txt', '\\1', file)
    data <- read.table(file, header = TRUE, sep = '\t')
    return(list(sample=sample,
                read_count=data$sumWeightAfter,
                umi_count=data$Total_uniqueMoleculeCount))
}) %>% 
  data.table::rbindlist()
```

### Large

```{r}
figures[['saturation_umis_large']] <- 
downsampling_umis %>%
  left_join(qc_metadata, by = 'sample') %>%
  filter(size=='L') %>%
  arrange(sample, read_count) %>%
  ggplot(aes(x=read_count, y=umi_count, color=individual, group=sample)) +
  geom_path() +
  # theme_bw() +
  scale_x_log10() +
  facet_wrap(~ kit, scales='free_y', ncol=1) +
  labs(x='Total reads', y='UMIs', color='Individual')
figures[['saturation_umis_large']] 
```

### Small

```{r}
figures[['saturation_umis_small']] <- 
downsampling_umis %>%
  left_join(qc_metadata, by = 'sample') %>%
  filter(size=='S') %>%
  arrange(sample, read_count) %>%
  ggplot(aes(x=read_count, y=umi_count, color=individual, group=sample)) +
  geom_path() +
  # theme_bw() +
  scale_x_log10() +
  facet_wrap(~ kit, scales='free_y', ncol=1) +
  labs(x='Total reads', y='UMIs', color='Individual')
figures[['saturation_umis_small']] 
```

## Clonotypes

```{r}
files <- dir(here('../processing/mixcr'), 
             'downsampling_summary_[0-9]*.txt', 
             recursive = TRUE, 
             full.names = TRUE)
downsampling_clonotypes <- lapply(
  files, 
  function(file) {
    sample <- gsub('.*(.-F[0-9][^/]+).*', '\\1', file)
    size <- gsub('.*_summary_([0-9]*).txt', '\\1', file)
    data <- read.table(file, header = TRUE, sep = '\t')
    return(list(sample=sample,
                read_count=data$sumWeightAfter,
                clonotype_count=data$nElementsAfter))
}) %>% 
  data.table::rbindlist()
```

### Large

```{r}
figures[['saturation_clonotypes_large']] <- 
downsampling_clonotypes %>%
  left_join(qc_metadata, by = 'sample') %>%
  filter(size=='L') %>%
  arrange(sample, read_count) %>%
  ggplot(aes(x=read_count, y=clonotype_count, color=individual, group=sample)) +
  geom_path() +
  # theme_bw() +
  scale_x_log10() +
  facet_wrap(~ kit, scales='free_y', ncol=1) +
  labs(x='Total reads', y='Clonotypes', color='Individual')
figures[['saturation_clonotypes_large']] 
```

### Small

```{r}
figures[['saturation_clonotypes_small']] <- 
downsampling_clonotypes %>%
  left_join(qc_metadata, by = 'sample') %>%
  filter(size=='S') %>%
  arrange(sample, read_count) %>%
  ggplot(aes(x=read_count, y=clonotype_count, color=individual, group=sample)) +
  geom_path() +
  # theme_bw() +
  scale_x_log10() +
  # theme(text=element_text(size=16,  family="sans")) +
  facet_wrap(~ kit, scales='free_y', ncol=1) +
  labs(x='Total reads', y='Clonotypes', color='Individual')
figures[['saturation_clonotypes_small']] 
```

## Old method (r downsampling clonotypes)

### Large

```{r, eval=FALSE}
downsampled_obj_large <- list()
for (j in names(repFilter(rep_data,
            .method = 'by.meta',
            list(chain = include('TRA', 'TRB'),
                 size = include('L')))$data)) {
  for (i in 10^seq(5,7,length.out=10)) {
    k <- sum(rep_data$data[[j]]$Clones)
    if (i < k) {
      downsampled_obj_large[[paste0(j, '@', as.character(i))]] <- repSample(rep_data$data[[j]],
                                                                            .method='downsample',
                                                                            .n = i)
    }
  }
  downsampled_obj_large[[paste0(j, '@', as.character(k))]] <- rep_data$data[[j]]
}
```

```{r, eval=FALSE}
plotdata <- repExplore(downsampled_obj_large, .method = c("volume")) %>%
  separate(Sample, c('Sample', 'libsize'), sep='@') %>%
  mutate(libsize= as.numeric(libsize)) %>%
  left_join(rep_data$meta, by = 'Sample')
figures[['clone_saturation_curves_large']] <-
  ggplot(plotdata, aes(x=libsize, y=Volume, color=individual, group=Sample)) +
  geom_path() +
  geom_point(size=0.25) +
  theme_bw() +
  facet_nested(rows=vars(kit), cols = vars(chain)) +
  scale_x_log10() +
  theme(panel.spacing = unit(0, "lines")) +
  labs(x='UMI count', y='Unique clones', caption = 'The final point for each curve is the full library size')
figures[['clone_saturation_curves_large']]
```

# Render report

```{r}
## This only works if all the variables called in the format file 
## are in the current environment. Check the format file to tweak aesthetics 
rmarkdown::render(here('scripts/01-pipeline_QC.format.Rmd'),
                  output_file = '01_QC.html',
                  output_dir = here('reports'))
```